### ISSUE TITLE
Weights are revived after pruning due to the momentum.

### ISSUE THREAD

**Issue by sseung0703** [he-y/filter-pruning-geometric-median#47]
this code only masks the gradient, but not momentum.
So masked weights revived after pruning and they will affect to other weights.
Did you check it?

**Comment by he-y** [comment]
I didn't see the weight revived during training.
The accuracy before and after pruning is constant, as shown in this figure:
![image](https://user-images.githubusercontent.com/10541284/70706782-9d5d2c00-1d11-11ea-8bc1-af9ce89b35bf.png)

The pruning operation at each epoch is also double insurance to keep the filters be zero.

**Comment by sseung0703** [comment]
Yes because you apply pruning before training, which is for "soft blar bar" **your previous work**.
So, they didn't produce momentum. but if you remove them, momentum will revive.
I check it using your code.

**Comment by sseung0703** [comment]
anyway I don't want this issue is closed because it is not solved

**Comment by he-y** [comment]
I am not quite sure about your question.
You want the pruned filters revive or not?

**Comment by sseung0703** [comment]
It is not my needs. it should not revive because they are pruned. but your pruned weights are revived. 
It may be not affect so much because they have very small value. But in some cases it will be. 
And I think it is not your intentions. Pruned weights should be zero in all cases, am I right?

**Comment by Allhailankurgupta** [comment]
After pruning for resnet when we check the saved model the weights are not zero. They are reviving. Can you please tell how to solve the problem.
The weights should remain zero as they are pruned.

**Comment by Allhailankurgupta** [comment]
> this code only masks the gradient, but not momentum.
> So masked weights revived after pruning and they will affect to other weights.
> Did you check it?

Were you able to solve this? Can you please tell how to solve this.

**Comment by he-y** [comment]
@Allhailankurgupta 
Have you checked the released model at [google driver](https://drive.google.com/drive/u/1/folders/1w_Max8L5ICJZSrlha8UybHfICik-iX95).

I have just checked them and find the weights are not reviving.
![image](https://user-images.githubusercontent.com/10541284/75758179-c11b9580-5d87-11ea-804a-8f081cb51070.png)

Can you provide more details about how it reviving?

**Comment by sseung0703** [comment]
@he-y Yes, in your implementation, the filters are not revived, because Your codes kill filters before start training, so there are no gradients. However it is not same with your paper.
If you make your code same with your paper, they will be revived.

@Allhailankurgupta the author's code and paper have errors, so you should fix yourself. To prevent reviving, you should mask gradient also.

**Comment by he-y** [comment]
@sseung0703 
I finally understand what you are talking about.

Yes, I kill filters before start training / retraining, because this is what a conventional **three stages** pruning does.
The three stages consist of training, pruning and retraining. You can take a look at: _Learning both Weights and Connections for Efficient Neural Networks_

I don't write this in detail in my paper as it's a kind of common-sense in pruning papers. 
Thanks for your question and I will modify the paper to make it clear.

**Comment by sseung0703** [comment]
@he-y No, you write this in detail in your paper.
![image](https://user-images.githubusercontent.com/26036843/75761139-e4d8de80-5d7b-11ea-9645-a07ffc817625.png)

In your algorithm in paper, you update pre-trained network In line 4 and prune network in line 7.

However in line 212-215 of pruning_cifar10.py , you prune network before training.

**Comment by sseung0703** [comment]
@Allhailankurgupta Also, if you want to use this code for your paper, I recommend you to reproduce all the results. I failed to got same results with paper using this codes. I think they are also fake.

**Comment by Allhailankurgupta** [comment]
> @Allhailankurgupta Also, if you want to use this code for your paper, I recommend you to reproduce all the results. I failed to got same results with paper using this codes. I think they are also fake.

so should I try to implement the code by myself from scratch

**Comment by sseung0703** [comment]
@Allhailankurgupta No, you shouldn't. Just implement new step function which doesn't update masked filters. But, the easiest way is applying "do_mask" and "do_simmilar_mask " after all step.

But, actually, I recommend re-implementing this algorithm. because there are so many errors in this repo.

**Comment by he-y** [comment]
Both the code and the algorithm in the paper would lead to a model with zeros, which means the effect of pruning.
